crates/data-structures/src/slim_slice.rs

```diff
@@ -0,0 +1,1801 @@
+    #[inline]
+    fn from(slice: SlimSliceMut<'_, T>) -> Self {
+        (&slice).into()
+    }
+}
+impl<T: Clone> From<SlimSliceMut<'_, T>> for Box<[T]> {
+    #[inline]
+    fn from(slice: SlimSliceMut<'_, T>) -> Self {
+        slice.deref().into()
+    }
+}
+impl<T: Clone> From<SlimSliceMut<'_, T>> for Vec<T> {
+    #[inline]
+    fn from(slice: SlimSliceMut<'_, T>) -> Self {
+        slice.deref().into()
+    }
+}
+
+impl<'a, T> TryFrom<&'a mut [T]> for SlimSliceMut<'a, T> {
+    type Error = LenTooLong<&'a mut [T]>;
+
+    #[inline]
+    fn try_from(slice: &'a mut [T]) -> Result<Self, Self::Error> {
+        ensure_len_fits!(slice);
+        // SAFETY: ^-- satisfies `len <= u32::MAX`.
+        Ok(unsafe { Self::from_slice_unchecked(slice) })
+    }
+}
+
+/// Converts `&mut [T]` into the slim limited version.
+///
+/// Panics when `slice.len() > u32::MAX`.
+#[inline]
+pub fn slice_mut<T>(s: &mut [T]) -> SlimSliceMut<'_, T> {
+    expect_fit(s)
+}
+
+// =============================================================================
+// Shared string slice reference
+// =============================================================================
+
+/// A shared reference to `str` limited to `u32::MAX` in length.
+#[repr(transparent)]
+#[derive(Clone, Copy)]
+pub struct SlimStr<'a> {
+    /// The raw byte slice.
+    raw: SlimSlice<'a, u8>,
+}
+
+impl<'a> SlimStr<'a> {
+    /// Converts `s` to `Self` without checking `s.len() <= u32::MAX`.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that `s.len() <= u32::MAX`.
+    #[inline]
+    const unsafe fn from_str_unchecked(s: &'a str) -> Self {
+        // SAFETY: Caller has promised that `s.len() <= u32::MAX`.
+        let raw = unsafe { SlimSlice::from_slice_unchecked(s.as_bytes()) };
+        // SAFETY: `s: &str` is always UTF-8.
+        Self { raw }
+    }
+}
+
+impl Deref for SlimStr<'_> {
+    type Target = str;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        // SAFETY: Data is derived from `str` originally so it's valid UTF-8.
+        unsafe { from_utf8_unchecked(self.raw.deref()) }
+    }
+}
+
+impl Debug for SlimStr<'_> {
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Debug::fmt(self.deref(), f)
+    }
+}
+
+impl Display for SlimStr<'_> {
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        Display::fmt(self.deref(), f)
+    }
+}
+
+impl Hash for SlimStr<'_> {
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        Hash::hash(self.deref(), state)
+    }
+}
+
+impl Eq for SlimStr<'_> {}
+impl PartialEq for SlimStr<'_> {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        self.deref() == other.deref()
+    }
+}
+impl PartialEq<str> for SlimStr<'_> {
+    #[inline]
+    fn eq(&self, other: &str) -> bool {
+        self.deref() == other
+    }
+}
+
+impl Ord for SlimStr<'_> {
```

Comment:
It's hard to say without knowing how all this is used, but the provided set of impls feels a bit ad hoc. I suppose it can change later.

eg why not `impl Ord for SlimSlice<T> where T: Ord` etc.